solr/contrib/ltr/src/java/org/apache/solr/ltr/ranking/ModelQuery.java

```
@@ -172,48 +181,124 @@ private boolean equalsTo(ModelQuery other) {
+      this.req = req;
+    }
+
+    public FeatureWeight call() throws Exception{
+      try {
+        FeatureWeight fw  = f.createWeight(searcher, needsScores, req, originalQuery, efi);
+        return fw;
+      } catch (final Exception se) {
+        throw new FeatureException("Exception from createWeight for " + f.toString() + " "
+            + se.getMessage(), se);
+      } finally {
+        querySemaphore.release();
+        LTRThreadModule.ltrSemaphore.release();
+      }
+    }
+  } // end of call CreateWeightCallable
+
+  private void createWeightsParallel(IndexSearcher searcher, boolean needsScores, float boost,
+      List<FeatureWeight > featureWeights, Collection<Feature> features) throws IOException {
+
+    final SolrQueryRequest req = getRequest();
+
+    Executor executor = LTRThreadModule.createWeightScoreExecutor;
+    if  (LTRThreadModule.ltrSemaphore == null ){
+      LTRThreadModule.ltrSemaphore = new Semaphore(LTRThreadModule.getMaxThreads());
+    }
+    List<Future<FeatureWeight> > futures = new ArrayList<>(features.size());
+    try{
+      for (final Feature f : features) {
+        CreateWeightCallable callable = new CreateWeightCallable(f, searcher, needsScores, req);
+        RunnableFuture<FeatureWeight> runnableFuture = new FutureTask<>(callable);
+        querySemaphore.acquire(); // always acquire before the ltrSemaphore is acquired, to guarantee a that the current query is within the limit for max. threads 
+        LTRThreadModule.ltrSemaphore.acquire();//may block and/or interrupt
+        
+        executor.execute(runnableFuture);//releases semaphore when done
+        futures.add(runnableFuture);
+      }
+      //Loop over futures to get the feature weight objects
+      for (final Future<FeatureWeight> future : futures) {
+        featureWeights.add(future.get());
+      }
+    } catch (InterruptedException e) {
+      log.info("Error while creating weights in LTR: InterruptedException", e);
```

Should we be silently failing for these exceptions?